File: prototypes/pdf-structure/probes/parse_table5.py
import re, json, sys

with open("table5_raw.txt", encoding="utf-8") as f:
    raw_lines = [l.rstrip("\n") for l in f]

NOISE_EXACT = {
    "Structure Type", "Children", "Parents", "Occ.",
    "Table 5 (continued)",
    "Table 5 — Parent-child relationships between the PDF 1.7 elements and PDF 2.0 elements",
    "© ISO 2023 – All rights reserved\t",
    "© ISO 2023 – All rights reserved",
    "ISO/TS 32005:2023(E)",
    "Sold by the PDF Association to Rabiullah Lotfullah  20978 | September 2, 2026 |",
    "Single user only, copying and networking prohibited.",
    "﻿",
    "c",  # stray OCR artifact observed twice mid-sequence (lines ~3794, ~4929 of table5_raw.txt)
}

OCC_SET = {"0..n", "1..n", "0..1", "1", "∅*", "‡", "[a]", "[b]"}

tokens = []
token_src_line = []  # 1-based line number in table5_raw.txt, for traceability
dropped_digit_lines = []

for i, l in enumerate(raw_lines, start=1):
    s = l.strip()
    if s == "":
        continue
    if re.match(r"^-{5,} PAGE \d+ -{5,}$", s):
        continue
    if s in NOISE_EXACT:
        continue
    if re.match(r"^\d+$", s) and s != "1":
        dropped_digit_lines.append((i, s))
        continue
    tokens.append(s)
    token_src_line.append(i)

print(f"Total data tokens after noise filtering: {len(tokens)}", file=sys.stderr)
print(f"Dropped bare page-number digit lines: {len(dropped_digit_lines)}", file=sys.stderr)

rows = []
pos = 0
n_tok = len(tokens)
errors = []

while pos < n_tok:
    header = tokens[pos]
    header_line = token_src_line[pos]
    if header in OCC_SET:
        errors.append((pos, token_src_line[pos], f"Expected row header (type name), got OCC token {header!r}"))
        pos += 1
        continue
    pos += 1

    children_occ = []
    while pos < n_tok and tokens[pos] in OCC_SET:
        children_occ.append(tokens[pos])
        pos += 1
    n = len(children_occ)
    children_types = tokens[pos:pos+n]
    child_lines = token_src_line[pos:pos+n]
    pos += n

    parents_occ = []
    while pos < n_tok and tokens[pos] in OCC_SET:
        parents_occ.append(tokens[pos])
        pos += 1
    m = len(parents_occ)
    parents_types = tokens[pos:pos+m]
    parent_lines = token_src_line[pos:pos+m]
    pos += m

    rows.append({
        "header": header,
        "header_line": header_line,
        "children": list(zip(children_types, children_occ, child_lines)),
        "parents": list(zip(parents_types, parents_occ, parent_lines)),
    })

print(f"Parsed {len(rows)} row-blocks", file=sys.stderr)
print(f"Parse errors: {len(errors)}", file=sys.stderr)
for e in errors:
    print("  ERROR:", e, file=sys.stderr)

# Sanity: list row headers in order with line numbers and counts
for r in rows:
    print(f"L{r['header_line']:5d}  {r['header']:16s} children={len(r['children']):3d} parents={len(r['parents']):3d}")

with open("table5_parsed.json", "w", encoding="utf-8") as f:
    json.dump(rows, f, ensure_ascii=False, indent=1)

File: prototypes/pdf-structure/probes/parse_table5.js
const fs = require('fs');

const rawLines = fs.readFileSync('table5_raw.txt', 'utf8').split(/\r?\n/);

const BOM = String.fromCharCode(0xFEFF);
const BS = String.fromCharCode(0x08);

const NOISE_EXACT = new Set([
  "Structure Type", "Children", "Parents", "Occ.",
  "Table 5 (continued)",
  "Table 5 — Parent-child relationships between the PDF 1.7 elements and PDF 2.0 elements",
  "© ISO 2023 – All rights reserved\t",
  "© ISO 2023 – All rights reserved",
  "ISO/TS 32005:2023(E)",
  "Sold by the PDF Association to Rabiullah Lotfullah  20978 | September 2, 2026 |",
  "Single user only, copying and networking prohibited.",
  BOM,
  "c",
]);

const OCC_SET = new Set(["0..n", "1..n", "0..1", "1", "∅*", "‡", "[a]", "[b]"]);

const ctrlOnlyRe = new RegExp("^[" + BOM + BS + "]*$");

const tokens = [];
const tokenSrcLine = [];
const droppedDigits = [];

rawLines.forEach((l, idx) => {
  const lineNo = idx + 1;
  const s = l.trim();
  if (s === "") return;
  if (ctrlOnlyRe.test(s)) return; // BOM / stray backspace control-char artifacts
  if (/^-{5,} PAGE \d+ -{5,}$/.test(s)) return;
  if (NOISE_EXACT.has(s)) return;
  if (/^\d+$/.test(s) && s !== "1") {
    droppedDigits.push([lineNo, s]);
    return;
  }
  tokens.push(s);
  tokenSrcLine.push(lineNo);
});

console.error(`Total data tokens after noise filtering: ${tokens.length}`);
console.error(`Dropped bare page-number digit lines: ${droppedDigits.length}`);

const rows = [];
let pos = 0;
const n_tok = tokens.length;
const errors = [];

while (pos < n_tok) {
  const header = tokens[pos];
  const headerLine = tokenSrcLine[pos];
  if (OCC_SET.has(header)) {
    errors.push([pos, tokenSrcLine[pos], `Expected row header, got OCC token ${header}`]);
    pos += 1;
    continue;
  }
  pos += 1;

  const childrenOcc = [];
  while (pos < n_tok && OCC_SET.has(tokens[pos])) {
    childrenOcc.push(tokens[pos]);
    pos += 1;
  }
  const n = childrenOcc.length;
  const childrenTypes = tokens.slice(pos, pos + n);
  const childLines = tokenSrcLine.slice(pos, pos + n);
  pos += n;

  const parentsOcc = [];
  while (pos < n_tok && OCC_SET.has(tokens[pos])) {
    parentsOcc.push(tokens[pos]);
    pos += 1;
  }
  const m = parentsOcc.length;
  const parentsTypes = tokens.slice(pos, pos + m);
  const parentLines = tokenSrcLine.slice(pos, pos + m);
  pos += m;

  rows.push({
    header, headerLine,
    children: childrenTypes.map((t, i) => [t, childrenOcc[i], childLines[i]]),
    parents: parentsTypes.map((t, i) => [t, parentsOcc[i], parentLines[i]]),
  });
}

console.error(`Parsed ${rows.length} row-blocks`);
console.error(`Parse errors: ${errors.length}`);
errors.forEach(e => console.error("  ERROR:", JSON.stringify(e)));

rows.forEach(r => {
  console.log(`L${String(r.headerLine).padStart(5)}  ${r.header.padEnd(16)} children=${String(r.children.length).padStart(3)} parents=${String(r.parents.length).padStart(3)}`);
});

fs.writeFileSync('table5_parsed.json', JSON.stringify(rows, null, 1), 'utf8');
